### ISSUE TITLE
Explore GSD file format for ESPResSo

### ISSUE THREAD

**Issue by jngrad** [espressomd/espresso#4793]
**TL;DR:** GSD shares a lot of similarities with hdf5, and the HOOMD GSD schema is essentially a subset of the ESPResSo H5MD schema. The hdf5/H5MD implementation in ESPResSo has a lot of overhead. GSD could provide a better alternative for HPC.

## Background

The [GSD](https://github.com/glotzerlab/gsd) file format writes labeled time series  as "chunks" (e.g. particle positions, particle types, box length), which can then later be retrieved either as time series (e.g. full numy array), or as frames, which provide slices with e.g. the box length and all particle positions at time *t*. While the chunks can be written in free form, the [HOOMD Schema](https://gsd.readthedocs.io/en/v3.1.1/schema-hoomd.html#data-chunks) provides a list of particle properties with shape information that overlaps pretty well with the ESPResSo H5MD schema. Only a few properties are missing, such as the dipole moment or the Lees-Edwards offset.

The GSD file format is designed with HPC in mind, and is supported by [MDAnalysis](https://docs.mdanalysis.org/stable/index.html) (see [GSD chapter](https://docs.mdanalysis.org/stable/documentation_pages/coordinates/GSD.html)), [MDTraj](https://github.com/mdtraj/mdtraj) (see [GSD module](https://github.com/mdtraj/mdtraj/blob/1.9.9/mdtraj/formats/gsd.py)), [freud](https://freud.readthedocs.io/en/latest/) (see [GSD section](https://freud.readthedocs.io/en/latest/topics/datainputs.html#gsd-trajectories)), and VMD via the plugin [gsd-vmd](https://github.com/mphowardlab/gsd-vmd). Whether their parser is specific to the HOOMD GSD schema is unclear. If it is, we might have to either create an ESPResSo schema that derives from the HOOMD GSD schema, or create an independent ESPResSo GSD schema and contribute a parser to all these tools.

## Open questions

The drawback of deriving our schema from the HOOMD GSD schema is that the ESPResSo H5MD schema was designed such that accessing a time series that is missing from the hdf5 file dataset throws an error, whereas the HOOMD GSD schema populates these missing time series by a default value. The latter can be problematic for automated data analysis pipelines, and would require users to always verify time series are not populated with a default value before running an analysis method.

We also need to find a way to store a unit system. The H5MD specifications allow storing very basic information as a string containing the SI unit with a scaling factor (see [Units chapter](https://www.nongnu.org/h5md/modules/units.html)). The HOOMD GSD schema doesn't offer a way to store unit information.

While exploring this data format, the following points need to be addressed:
* do MDTraj/MDAnalysis/freud/gsd-vmd rely on the HOOMD GSD schema, or do they work out-of-the-box with a custom GSD file with the same chunk names?
* how to store a unit system in the GSD file? Possibly as metadata for the first frame of each data series.
* what is the performance profile of GSD vs. hdf5? Use the benchmark files in [talk/2023-grad-fairmat-mainz](https://gitlab.icp.uni-stuttgart.de/talk/2023-grad-fairmat-mainz/-/tree/main/figures/benchmarks?ref_type=heads) on a SSD drive, once a C++ prototype is available in ESPResSo

The first two points can be answered with only python. The third one only needs a basic C++ implementation that supports only particle positions and types. A full C++ re-implementation of the ESPResSo H5MD schema can then be carried out using the H5MD core I/O C++ source files and the corresponding Python unit tests as a reference.

## Example

To get started, run `python3 -m pip install gsd` and run the following:

```python
import gsd.fl
import gsd.hoomd
import numpy as np
np.random.seed(42)

print("Writing HOOMD GSD file")
with gsd.hoomd.open(name="file.gsd", mode="w") as f:
    frame = gsd.hoomd.Frame()
    frame.configuration.step = 0
    frame.configuration.box = [1., 2., 3., 0., 0., 0.]
    frame.particles.N = 4
    frame.particles.position = np.random.random(size=(4, 3))
    frame.particles.typeid = [1, 2, 2, 2]
    f.extend([frame])
    frame = gsd.hoomd.Frame()
    frame.configuration.step = 1
    frame.configuration.box = [1., 2., 2., 0., 0., 0.]
    frame.particles.N = 3
    frame.particles.position = np.random.random(size=(3, 3))
    frame.particles.typeid = [1, 2, 2]
    f.extend([frame])

print("Opening HOOMD GSD file")
with gsd.hoomd.open(name="file.gsd", mode="r") as f:
    for frame in f:
    	print(frame.particles.typeid)

print("Writing custom GSD file")
with gsd.fl.open(name="file.gsd", mode="w", application="ESPResSo",
                 schema="ESPResSo", schema_version=[1, 0]) as f:
    f.write_chunk(name='chunk1', data=np.array([1,2,3,4], dtype=np.float32))
    f.write_chunk(name='chunk2', data=np.array([[5,6],[7,8]], dtype=np.float32))
    f.end_frame()
    f.write_chunk(name='chunk1', data=np.array([1,2,3,4], dtype=np.float32))
    f.write_chunk(name='chunk2', data=np.array([[5,6],[7,8]], dtype=np.float32))
    f.end_frame()

print("Opening custom GSD file")
with gsd.fl.open(name="file.gsd", mode="r") as f:
    data = f.read_chunk(frame=0, name='chunk1')
    print(data)
```

## Acknowledgments

This proposal is an outcome of the CECAM Flagship Workshop [FAIR and TRUE Data Processing for Soft Matter Simulations](https://www.cecam.org/workshop-details/1228), Mainz, September 2023. Thanks to @b-butler for the fruitful discussion.
